include/wayland/mir/wayland/wayland_base.h

```diff
@@ -37,12 +41,62 @@ class Resource
     };
 
     Resource();
-    virtual ~Resource() = default;
+    ~Resource();
 
     Resource(Resource const&) = delete;
     Resource& operator=(Resource const&) = delete;
+
+    auto destroyed_flag() const -> std::shared_ptr<bool>;
+
+private:
+    std::shared_ptr<bool> mutable destroyed;
+};
+
+/// A weak handle to a Wayland resource
+/// May only be safely used from the Wayland thread
+template<typename T>
+class Handle
```

Comment:
Umm. When I said "use a handle class" I didn't mean "use a class called "Handle"".

The interface and use is closer to the optional idiom than to a smart pointer or handle/body. But as this indirection is about lifetime, maybe "Ephemeral<>"?